File: sherry/week-5/453. Flatten Binary Tree to Linked List.java
//chanllagence-> don;t use pre-order
// use devide conter, save pre-order's last node != null, last node
// root.left = null, root.tight = left.lastnode left,rightnode = root.rightnode

/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: a TreeNode, the root of the binary tree
     * @return: nothing
     */
    public void flatten(TreeNode root) {
        // write your code here
        // root.l
        helper(root);
    }
    //the helper method returns the pointer to last node
    private TreeNode helper(TreeNode root) {
        if(root == null) {
            return null;
        }

        TreeNode leftLast = helper(root.left);
        TreeNode rightLast = helper(root.right);

        //put root.right = root.left
        //put leftlast.right = root.right;
        if(leftLast != null) {
            leftLast.right = root.right;
            root.right = root.left;
            //NOTE: left need points to null
            root.left = null;
        }

        if(rightLast != null) {
            return rightLast;
        }
        if (leftLast != null) {
            return leftLast;
        }
        return root;

    }
}

File: Joe Zheng/week 5/453. Flatten Binary Tree to Linked List.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: a TreeNode, the root of the binary tree
    @return: nothing
    """
    def flatten(self, root):
        # write your code here
        # create a stack
        if root is None:
            return
        
        self.flatten(root.left)
        self.flatten(root.right)
        
        tmp = root
        if tmp.left is None:
            return

        # left most        
        tmp = tmp.left
        # left tree last right
        while tmp.right:
            tmp = tmp.right
        # put root right to the current.right, but cause current is the next one
        tmp.right = root.right
        # switch from left to right and remove left
        root.right = root.left
        root.left = None
              
            
# {}
# {1,#,2,#,4,#,5,#,3,#,6}
# {1,#,2,4,#,#,5,#,3,#,6}
# {1,#,2,4,#,7,5,#,3,#,6}
# {1,1,1,1,1,1,1,1,1,1,1,1,1,1,1}
# {98,97,#,88,#,84,#,79,87,64,#,#,#,63,69,62,#,#,#,30,#,27,59,9,#,#,#,3,#,0,#,-4,#,-16,#,-18,-7,-19,#,#,#,-23,#,-34,#,-42,#,-59,#,-63,#,-64,#,-69,#,-75,#,-81}
